src/main/java/appeng/recipes/factories/recipes/PartRecipeFactory.java

```diff
@@ -45,25 +60,149 @@ public IRecipe parse( JsonContext context, JsonObject json )
 		else
 		{
 			throw new JsonSyntaxException( "Applied Energistics 2 was given a custom recipe that it does not know how to handle!\n" + "Type should either be '" + AppEng.MOD_ID + ":shapeless' or '" + AppEng.MOD_ID + ":shaped', got '" + type + "'!" );
 		}
 	}
+
+	// Copied from ShapedOreRecipe.java, modified a bit.
+	private static ShapedOreRecipe shapedFactory( JsonContext context, JsonObject json )
+	{
+		String group = JsonUtils.getString( json, "group", AppEng.MOD_ID + ":part_shaped_crafting" );
+
+		Map<Character, Ingredient> ingMap = Maps.newHashMap();
+		for( Map.Entry<String, JsonElement> entry : JsonUtils.getJsonObject( json, "key" ).entrySet() )
+		{
+			if( entry.getKey().length() != 1 )
+			{
+				throw new JsonSyntaxException( "Invalid key entry: '" + entry.getKey() + "' is an invalid symbol (must be 1 character only)." );
+			}
+			if( " ".equals( entry.getKey() ) )
+			{
+				throw new JsonSyntaxException( "Invalid key entry: ' ' is a reserved symbol." );
+			}
+
+			ingMap.put( entry.getKey().toCharArray()[0], CraftingHelper.getIngredient( entry.getValue(), context ) );
+		}
+
+		ingMap.put( ' ', net.minecraft.item.crafting.Ingredient.EMPTY );
+
+		JsonArray patternJ = JsonUtils.getJsonArray( json, "pattern" );
+
+		if( patternJ.size() == 0 )
+		{
+			throw new JsonSyntaxException( "Invalid pattern: empty pattern not allowed" );
+		}
+
+		String[] pattern = new String[patternJ.size()];
+		for( int x = 0; x < pattern.length; ++x )
+		{
+			String line = JsonUtils.getString( patternJ.get( x ), "pattern[" + x + "]" );
+			if( x > 0 && pattern[0].length() != line.length() )
+			{
+				throw new JsonSyntaxException( "Invalid pattern: each row must  be the same width" );
+			}
+			pattern[x] = line;
+		}
+
+		CraftingHelper.ShapedPrimer primer = new CraftingHelper.ShapedPrimer();
+		primer.width = pattern[0].length();
+		primer.height = pattern.length;
+		primer.mirrored = JsonUtils.getBoolean( json, "mirrored", true );
+		primer.input = NonNullList.withSize( primer.width * primer.height, net.minecraft.item.crafting.Ingredient.EMPTY );
+
+		Set<Character> keys = Sets.newHashSet( ingMap.keySet() );
+		keys.remove( ' ' );
+
+		int x = 0;
+		for( String line : pattern )
+		{
+			for( char chr : line.toCharArray() )
+			{
+				net.minecraft.item.crafting.Ingredient ing = ingMap.get( chr );
+				if( ing == null )
+				{
+					throw new JsonSyntaxException( "Pattern references symbol '" + chr + "' but it's not defined in the key" );
+				}
+				primer.input.set( x++, ing );
+				keys.remove( chr );
+			}
+		}
+
+		if( !keys.isEmpty() )
+		{
+			throw new JsonSyntaxException( "Key defines symbols that aren't used in pattern: " + keys );
+		}
+
+		JsonObject resultObject = (JsonObject) json.get( "result" );
+		int count = JsonUtils.getInt( resultObject, "count", 1 );
+
+		String ingredient = resultObject.get( "part" ).getAsString();
+		Object result = (Object) Api.INSTANCE.registries().recipes().resolveItem( AppEng.MOD_ID, ingredient );
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Remove this unnecessary cast to "Object". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AS1905)